File: solutions/haskell/rotational-cipher/1/src/RotationalCipher.hs
{-# LANGUAGE ImportQualifiedPost #-}

module RotationalCipher (rotate) where

import Data.Char qualified as C

rotate :: Int -> String -> String
rotate n =
    let
        calc amt ch base =
            C.chr $ C.ord base + (C.ord ch - C.ord base + amt) `rem` 26
        rot nn ch
            | C.isUpper ch = calc nn ch 'A'
            | C.isLower ch = calc nn ch 'a'
            | otherwise = ch
     in
        map (rot n)

File: solutions/ruby/rotational-cipher/1/rotational_cipher.rb
# frozen_string_literal: true

module RotationalCipher
  def self.rotate(text, rot)
    rotate_char = ->(c, rot, a_ord) { ((c.ord - a_ord + rot) % 26 + a_ord).chr }
    text.chars.map do |c|
      ('a'..'z').include?(c) ? rotate_char.(c, rot, 'a'.ord) :
        ('A'..'Z').include?(c) ? rotate_char.(c, rot, 'A'.ord) :
        c
    end.join('')
  end
end
